moonraker/components/authorization.py

```diff
@@ -356,16 +388,71 @@ async def _handle_password_reset(self,
             'action': "user_password_reset"
         }
 
+    def _login_ldap_user(self, username, password) -> bool:
+        base_dn = str(self.ldap_base_dn)
+        client = LDAPClient(self._generate_ldap_url_(str(self.ldap_server)))
+        client.set_credentials("SIMPLE", self.ldap_bind_dn,
+                               self.ldap_bind_password)
+        client.set_cert_policy("allow")
+        conn = LDAPConnection
+        try:
+            conn = client.connect(is_async=False, timeout=10)
+            ldap_filter = ("(&(objectClass=Person)(%s=" + '%s))') % \
+                          ("sAMAccountName"
+                           if self.ldap_type_ad
+                           else "uid",
+                           username)
+            user = conn.search(
+                base_dn,
+                LDAPSearchScope.SUBTREE,
+                ldap_filter,
+                ['memberOf']
+            )
+            auth_username = str(user[0]['DN'])
+            client.set_credentials("SIMPLE", auth_username, password)
+            conn = client.connect(is_async=False, timeout=10)
+            conn.close()
+            if self.ldap_group_dn is None:
+                return True
+            if len(user[0]['memberOf']) > 0:
+                for group in user[0]['memberOf']:
+                    if str(group) == str(self.ldap_group_dn):
+                        return True
+        except ConnectionError:
+            return False
+        except AuthenticationError:
+            conn.close()
+            return False
+        except TimeoutError:
+            return False
+        conn.close()
```

Comment:
Given that `conn.close()`is always called I would recommend using a context manager.  All 3 bonsai errors can be caught in the same statement.  Also I would recommend raising the authentication error here rather than return a value.  For example:

```python
try:
    async with client.connect(is_async=True, timeout=10) as conn:
         # do ldap login
except (ConnectionError, AuthenticationError, TimeoutError) as e:
    raise self.server.error("LDAP login error", 401) from e
raise self.server.error("ldap: Invalid Username or Password", 401)
```

Reply:
Hi, I wanted to ask what you mean by the context manager, I have not yet fully understood. Maybe the current code fits you already.